File: ios/TourGraph/TourGraph/TourGraph/Models/Chain.swift
import Foundation
import GRDB

/// Raw row from the `six_degrees_chains` table.
struct ChainRow: Identifiable, Codable, FetchableRecord, Sendable {
    let id: Int
    let cityFrom: String
    let cityTo: String
    let chainJson: String
    let generatedAt: String?

    enum CodingKeys: String, CodingKey, ColumnExpression {
        case id
        case cityFrom = "city_from"
        case cityTo = "city_to"
        case chainJson = "chain_json"
        case generatedAt = "generated_at"
    }
}

/// One link in a Six Degrees chain.
struct ChainLink: Codable, Sendable {
    let city: String
    let country: String
    let tourTitle: String
    let tourId: Int?
    let connectionToNext: String?
    let theme: String

    enum CodingKeys: String, CodingKey {
        case city, country, theme
        case tourTitle = "tour_title"
        case tourId = "tour_id"
        case connectionToNext = "connection_to_next"
    }
}

/// Parsed chain data from the JSON blob.
struct ChainData: Codable, Sendable {
    let cityFrom: String
    let cityTo: String
    let chain: [ChainLink]
    let summary: String

    enum CodingKeys: String, CodingKey {
        case cityFrom = "city_from"
        case cityTo = "city_to"
        case chain, summary
    }
}

/// Full chain with metadata, ready for display.
struct Chain: Identifiable, Sendable {
    let id: Int
    let cityFrom: String
    let cityTo: String
    let summary: String
    let links: [ChainLink]
    let slug: String
    let generatedAt: String?

    init(row: ChainRow) {
        self.id = row.id
        self.generatedAt = row.generatedAt

        let data = (try? JSONDecoder().decode(ChainData.self, from: Data(row.chainJson.utf8)))
        self.cityFrom = data?.cityFrom ?? row.cityFrom
        self.cityTo = data?.cityTo ?? row.cityTo
        self.summary = data?.summary ?? ""
        self.links = data?.chain ?? []

        let slugText = "\(self.cityFrom)-\(self.cityTo)"
            .lowercased()
            .replacingOccurrences(of: " ", with: "-")
        self.slug = slugText.filter { $0.isLetter || $0.isNumber || $0 == "-" }
    }
}

File: android/TourGraph/app/src/main/java/com/nikhilsi/tourgraph/model/Chain.kt
package com.nikhilsi.tourgraph.model

import android.database.Cursor
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class ChainLink(
    val city: String,
    val country: String,
    @SerialName("tour_title") val tourTitle: String,
    @SerialName("tour_id") val tourId: Int,
    @SerialName("connection_to_next") val connectionToNext: String? = null,
    val theme: String? = null
)

@Serializable
data class ChainData(
    @SerialName("city_from") val cityFrom: String,
    @SerialName("city_to") val cityTo: String,
    val chain: List<ChainLink>,
    val summary: String
)

private val chainJson = kotlinx.serialization.json.Json { ignoreUnknownKeys = true }

data class Chain(
    val id: Int,
    val cityFrom: String,
    val cityTo: String,
    val chainData: ChainData,
    val slug: String?,
    val generatedAt: String?
) {
    val links: List<ChainLink> get() = chainData.chain
    val summary: String get() = chainData.summary

    companion object {
        fun fromCursor(cursor: Cursor): Chain? {
            val json = cursor.getString(cursor.getColumnIndexOrThrow("chain_json")) ?: return null
            val chainData = try {
                chainJson.decodeFromString<ChainData>(json)
            } catch (_: Exception) {
                return null
            }
            return Chain(
                id = cursor.getInt(cursor.getColumnIndexOrThrow("id")),
                cityFrom = cursor.getString(cursor.getColumnIndexOrThrow("city_from")) ?: "",
                cityTo = cursor.getString(cursor.getColumnIndexOrThrow("city_to")) ?: "",
                chainData = chainData,
                slug = cursor.getStringOrNull("slug"),
                generatedAt = cursor.getStringOrNull("generated_at")
            )
        }

        private fun Cursor.getStringOrNull(column: String): String? {
            val idx = getColumnIndex(column)
            return if (idx >= 0 && !isNull(idx)) getString(idx) else null
        }
    }
}
